core/src/apps/bitcoin/ownership.py

```diff
@@ -0,0 +1,136 @@
+from trezor import wire
+from trezor.crypto import bip32, hashlib, hmac
+
+from apps.bitcoin.multisig import multisig_pubkey_index
+from apps.bitcoin.writers import write_bytes_prefixed
+from apps.common import seed
+from apps.common.writers import (
+    empty_bytearray,
+    write_bitcoin_varint,
+    write_bytes_fixed,
+    write_uint8,
+)
+
+from . import common, scripts
+from .verification import SignatureVerifier
+
+if False:
+    from typing import List, Optional, Tuple
+    from trezor.messages.MultisigRedeemScriptType import MultisigRedeemScriptType
+    from trezor.messages.TxInputType import EnumTypeInputScriptType
+    from apps.common.coininfo import CoinInfo
+
+# This module implements the SLIP-0019 proof of ownership format.
+
+_VERSION_MAGIC = b"SL\x00\x19"
+_FLAG_USER_CONFIRMED = 0x01
+_OWNERSHIP_ID_LEN = 32
+
+
+def generate_proof(
+    node: bip32.HDNode,
+    script_type: EnumTypeInputScriptType,
+    multisig: MultisigRedeemScriptType,
+    coin: CoinInfo,
+    user_confirmed: bool,
+    ownership_ids: List[bytes],
+    script_pubkey: bytes,
+    commitment_data: Optional[bytes],
+) -> Tuple[bytes, bytes]:
+    flags = 0
+    if user_confirmed:
+        flags |= _FLAG_USER_CONFIRMED
+
+    proof = empty_bytearray(4 + 1 + 1 + len(ownership_ids) * _OWNERSHIP_ID_LEN)
```

Comment:
The end result is the proof.

Reply:
My idea is to
  * rename `proof` to `proof_body`,
  * make the function `write_bip322_signature_proof` return the pair (or concatenation) of `witness` and `script_sig` and rename it accordingly,
  * return the concatenation of `proof_body`, `script_sig` and `witness` at the end of `generate_proof`.

Just suggesting since it's in my opinion better readable.